### PR TITLE
feat: add infra metering implementation with e2e test

### PR DESCRIPTION
**Pull request by xiaohan1202**
Notice: Need to change the Resource CRD to be compatible with this module.

### CODE REVIEW

File: controllers/infra/controllers/infraresource_controller.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	// List all infras
+	err := r.List(ctx, &infraList)
+	if err != nil {
+		return err
+	}
+	// sync resource used for each infra
+	for i := range infraList.Items {
+		infra := infraList.Items[i]
+		// if infra is not running, skip it
+		if !r.checkInfraStatus(infra) {
```
Comment (xiao-jay) [inline controllers/infra/controllers/infraresource_controller.go:119]: rename `checkInfraStatusRunning` is better

File: controllers/infra/controllers/infraresource_controller.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	return fmt.Sprintf("%s-%s-%v", infra.Namespace, infra.Name, seqID)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *InfraResourceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	const controllerName = "InfraResource-controller"
+	r.Logger = ctrl.Log.WithName(controllerName)
+	r.Logger.V(1).Info("reconcile InfraResource-controller")
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1.InfraResource{}).
```
Comment (xiao-jay) [inline controllers/infra/controllers/infraresource_controller.go:220]: can use this to ignore status update lead to reconcile.
```
For(&meteringv1.InfraResource{}, builder.WithPredicates(predicate.Or(predicate.GenerationChangedPredicate{}))).
```

**Comment by xiao-jay** [comment]
Could you please to write a e2e test ,test your infraresource-controller can be created and the Resource CR can be watch by metering.If you finish a comment,please reply a done to let me know.

File: controllers/infra/controllers/infraresource_controller.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+// InfraResourceReconciler reconciles a InfraResource object
+type InfraResourceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	logr.Logger
+	MeteringSystemNameSpace string
+}
+
+//+kubebuilder:rbac:groups=metering.sealos.io,resources=infraresources,verbs=get;list;watch;create;update;patch;delete
```
[the remark this answers is not in the document]
Comment (xiaohan1202) [reply controllers/infra/controllers/infraresource_controller.go]: > group is `groups=infra.sealos.io`?

done

File: controllers/infra/controllers/infraresource_controller.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+		if infraResource.Spec.Resources == nil {
+			infraResource.Spec.Resources = make(map[v1.ResourceName]meteringv1.ResourceInfo)
+		}
+		//TODO: update resource crd: change used type to int64 and add price parameter
+		//infraResource.Spec.Resources[v1.ResourceName(resourceName)] = meteringv1.ResourceInfo{
+		//	Used:      Used,
+		//	TimeStamp: time.Now().Unix(),
+		//	NameSpace: infra.Namespace,
+		//	Price:     price,
+		//}
```
[the remark this answers is not in the document]
Comment (xiaohan1202) [reply controllers/infra/controllers/infraresource_controller.go:204]: > resource yaml like this is better.
> 
> ```
> Kind: Resource
> name: infra-resource-seqID
> annotation:
> 	data: {{flavor:xxxx,disk:20G,type:gp3},{instanceid:xxxx,disk:20G,type:gp3}
> Spec:
> 	name:infra/CPU
> 		Used:16
> 		cost:12300
> 	name:infra/Memory
> 		Used:24
> 		cost:12300
> 	name:infra/disk
> 		Used:100G
> 		cost:10000
>         name: infra/flow
>   	       used：100G
>   	       cost：10000
> ```

done

File: controllers/infra/controllers/infraresource_controller.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+func GetResourceName(namespaceName, infraName string, resourceName string, seqID int64) string {
+	return fmt.Sprintf("%s-%s-%s-%s-%v", namespaceName, infrav1.InfraResourcePricePrefix, infraName, resourceName, seqID)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *InfraResourceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	const controllerName = "InfraResource-controller"
+	r.Logger = ctrl.Log.WithName(controllerName)
+	r.Logger.V(1).Info("reconcile InfraResource-controller")
+	r.MeteringSystemNameSpace = os.Getenv(meteringv1.METERINGNAMESPACEENV)
```
[the remark this answers is not in the document]
Comment (xiaohan1202) [reply controllers/infra/controllers/infraresource_controller.go]: > 你不一定有这个env吧，resource没必要一定放metering-system namespace

done

File: controllers/infra/controllers/infraresource_controller.go (hunk 2 above)
Comment (xiaohan1202) [reply controllers/infra/controllers/infraresource_controller.go]: > can use this to ignore status update lead to reconcile.
> 
> ```
> For(&meteringv1.InfraResource{}, builder.WithPredicates(predicate.Or(predicate.GenerationChangedPredicate{}))).
> ```

done

File: controllers/infra/PROJECT [hunk 6]
```diff
@@ -13,4 +13,13 @@ resources:
...
   path: github.com/labring/sealos/controllers/infra/api/v1
   version: v1
+- api:
+    crdVersion: v1
+    namespaced: true
+  controller: true
+  domain: sealos.io
+  group: infra
+  kind: InfraResource
+  path: github.com/labring/sealos/api/v1
```
Comment (xiao-jay) [inline controllers/infra/PROJECT:23]: the path is wrong

File: controllers/infra/config/samples/infra_v1_infraresource.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
+apiVersion: infra.sealos.io/v1
+kind: InfraResource
+metadata:
+  name: infraresource-sample
+  namespace: metering-system
+spec:
+  resourceName: infra
+  interval: 60
```
Comment (xiao-jay) [inline controllers/infra/config/samples/infra_v1_infraresource.yaml:8]: you should interval is 1 Minute?
And the price should change it

File: controllers/infra/PROJECT (hunk 6 above)
Comment (xiaohan1202) [reply controllers/infra/PROJECT]: > the path is wrong

done

File: controllers/infra/config/samples/infra_v1_infraresource.yaml (hunk 7 above)
Comment (xiaohan1202) [reply controllers/infra/config/samples/infra_v1_infraresource.yaml:8]: > you should interval is 1 Minute? And the price should change it

I've changed the cost calculation to divide 60

**Comment by xiaohan1202** [comment]
> @xiaohan1202 Could you please to write a e2e test ,test your infraresource-controller can be created and the Resource CR can be watch by metering.If you finish a comment,please reply a done to let me know.

I've tested if the resouce controller can generate the cr properly, here's the yaml, please check:
image

File: controllers/infra/api/v1/infraresource_types.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// InfraResourceSpec defines the desired state of InfraResource
+type InfraResourceSpec struct {
+	ResourceName string                                             `json:"resourceName,omitempty"`
+	Resources    map[v1.ResourceName]meteringcommonv1.ResourcePrice `json:"resources,omitempty"`
+
+	// update used resources every Interval minutes
+	//+kubebuilder:default=50
```
Comment (xiao-jay) [inline controllers/infra/api/v1/infraresource_types.go:31]:
```
//+kubebuilder:default=50
```
this can delete

**Review by fanux — APPROVED** [review]

File: controllers/infra/controllers/infraresource_controller.go [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+			Namespace: common.InfraSystemNamespace,
+			Name:      meteringcommonv1.ExtensionResourcePricePrefix + infrav1.InfraResourcePricePrefix,
+		},
+	}
+
+	r.Logger.Info("create or update extensionResourcePrice", "infraController name", infraController.Name, "infraController Resources", infraController.Spec.Resources)
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, extensionResourcesPrice, func() error {
+		extensionResourcesPrice.Spec.Resources = infraController.Spec.Resources
+		extensionResourcesPrice.Spec.ResourceName = infraController.Spec.ResourceName
+		return controllerutil.SetControllerReference(infraController, extensionResourcesPrice, r.Scheme)
```
Comment (xiao-jay) [inline controllers/infra/controllers/infraresource_controller.go:114]: 应该在extensionResourcesPrice 中写入你统计的资源的GVK，你这边应该是infras.infra.sealos.io
下面的例子
```
extensionResourcesPrice.Spec.GroupVersionKinds = meteringv1.DefaultPodResourceGVK
```

File: controllers/infra/testdata/e2e/infra_metering_test.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
+package e2e
```
Comment (xiao-jay) [inline controllers/infra/testdata/e2e/infra_metering_test.go:1]: 有无自己打包镜像然后运行测试，能否通过

File: controllers/infra/main.go [hunk 11]
```diff
@@ -45,6 +48,10 @@ func init() {
 	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
 
 	utilruntime.Must(infrav1.AddToScheme(scheme))
+
+	utilruntime.Must(meteringcommonv1.AddToScheme(scheme))
+
+	utilruntime.Must(meteringv1.AddToScheme(scheme))
```
Comment (xiao-jay) [inline controllers/infra/main.go:54]: 应该不需要吧？可以给出需要的地方吗

File: controllers/infra/config/samples/infra_v1_infraresource.yaml [hunk 12]
```diff
@@ -9,15 +9,15 @@ spec:
   resources:
     infra/CPU:
       unit: "1"
-      price: 670
+      price: 12
```
Comment (xiao-jay) [inline controllers/infra/config/samples/infra_v1_infraresource.yaml:12]: 在部署的时候最好还是精细一点吧，比如unit是10，price是112，这个误差有点大呀，menory也是

File: controllers/infra/config/samples/infra_v1_infraresource.yaml [hunk 13]
```diff
@@ -9,15 +9,15 @@ spec:
...
       describe: "cost of per cpu per hour"
 
     infra/Memory:
       unit: "1G"
```
Comment (xiao-jay) [inline controllers/infra/config/samples/infra_v1_infraresource.yaml:16]: 记得改成Gi，1G=1000000000，1Gi=2^30

**Review by xiao-jay — APPROVED** [review]
这些小问题改一下，其他没问题了

**Pull request merged**